src/Components/Components/src/Forms/InputComponents/InputBase.cs

```diff
@@ -0,0 +1,189 @@
+        /// <param name="validationErrorMessage">If the value could not be parsed, provides a validation error message.</param>
+        /// <returns>True if the value could be parsed; otherwise false.</returns>
+        protected virtual bool TryParseValueFromString(string value, out T result, out string validationErrorMessage)
+            => throw new NotImplementedException($"Components that inherit from {nameof(InputBase<T>)} must override " +
+                $"{nameof(TryParseValueFromString)} in order to use {nameof(CurrentValueAsString)}.");
+
+        /// <summary>
+        /// Gets a string that indicates the status of the field being edited. This will include
+        /// some combination of "modified", "valid", or "invalid", depending on the status of the field.
+        /// </summary>
+        protected string FieldClass
+            => EditContext.FieldClass(FieldIdentifier);
+
+        /// <inheritdoc />
+        public override Task SetParametersAsync(ParameterCollection parameters)
+        {
+            parameters.SetParameterProperties(this);
+
+            if (EditContext == null)
+            {
+                // This is the first run
+                // Could put this logic in OnInit, but its nice to avoid forcing people who override OnInit to call base.OnInit()
+
+                if (CascadedEditContext == null)
+                {
+                    throw new InvalidOperationException($"{GetType()} requires a cascading parameter " +
+                        $"of type {nameof(Forms.EditContext)}. For example, you can use {GetType().FullName} inside " +
+                        $"an {nameof(EditForm)}.");
+                }
+
+                if (ValueExpression == null)
+                {
+                    throw new InvalidOperationException($"{GetType()} requires a value for the 'ValueExpression' " +
+                        $"parameter. Normally this is provided automatically when using 'bind-Value'.");
+                }
+
+                EditContext = CascadedEditContext;
+                FieldIdentifier = FieldIdentifier.Create(ValueExpression);
+                _nullableUnderlyingType = Nullable.GetUnderlyingType(typeof(T));
+            }
+            else if (CascadedEditContext != EditContext)
+            {
+                // Not the first run
+
+                // We don't support changing EditContext because it's messy to be clearing up state and event
+                // handlers for the previous one, and there's no strong use case. If a strong use case
+                // emerges, we can consider changing this.
+                throw new InvalidOperationException($"{GetType()} does not support changing the " +
+                    $"{nameof(Forms.EditContext)} dynamically.");
+            }
+
+            // For derived components, retain the usual lifecycle with OnInit/OnParametersSet/etc.
+            return base.SetParametersAsync(ParameterCollection.Empty);
```

Comment:
Why is this better to put after our init logic? fragile base class problem? 

It looks like a gotcha that we have to pass empty here.

Reply:
It's because the base class will call the `OnInit` (etc.) methods, and we want that to happen *after* we've run the subclass's logic to populate `EditContext` etc., in case those lifecycle methods try to access `EditContext` etc.